docs/gitbook/usage/envoy-progressive-delivery.md

```diff
@@ -0,0 +1,309 @@
+# Envoy Canary Deployments
+
+This guide shows you how to use Envoy and Flagger to automate canary deployments.
+
+### Prerequisites
+
+Flagger requires a Kubernetes cluster **v1.11** or newer and Envoy paired with [Crossover](https://github.com/mumoshu/crossover) sidecar.
+
+Create a test namespace:
+
+```bash
+kubectl create ns test
+```
+
+Install Envoy along with the sidecar with Helm:
+
+```bash
+git clone https://github.com/mumoshu/crossover.git
+
+cd crossover
+
+helm upgrade --install envoy stable/envoy \
+  --namespace test \
+  -f example/values.yaml \
+  -f <(cat <<EOF
+services:
+  podinfo:
+    smi:
+      enabled: true
+    backends:
+      podinfo-primary:
+        port: 9898
+        weight: 50
+      podinfo-canary:
+        port: 9898
+        weight: 50
```

Comment:
The canary weight should be zero otherwise half the calls will fail after Flagger initializes podinfo. I think crossover shouldn’t require the backends to be provided at startup since Flagger creates them. Could crossover work by just supplying the SMI traffic split name?

Reply:
> The canary weight should be zero otherwise half the calls will fail after Flagger initializes podinfo

Good catch! Fixed it.

> I think crossover shouldn’t require the backends to be provided at startup since Flagger creates them. Could crossover work by just supplying the SMI traffic split name?

I found one crossover bug that caused startup failures when there was no trafficsplits so it's easier to use now.

But probably what you wanted to suggest was about omitting the blow below as a whole, right?

```
      podinfo-primary:
        port: 9898
        weight: 100
      podinfo-canary:
        port: 9898
        weight: 0
```

It won't work as, at least, you need `port: 9898` in the config due to SMI limitation. TrafficSplits clearly don't include ports. TrafficTarget does contain ports but it is for access control from source-to-destination, not from proxy-to-destionation.

It's definitely possible to enhance crossover to omit weights, so that it looks:

```
      podinfo-primary:
        port: 9898
      podinfo-canary:
        port: 9898
```

Crossover would implicitly treat the first backend `podinfo-primary` as weight=100 until it finds corresponding trafficsplit. If trafficsplit found, it would just use weights declared in it.

Does it help?